src/sdk/namespace_fs.js

```diff
@@ -1759,6 +1800,213 @@ class NamespaceFS {
+                    if (!deleted_delete_marker) deleted_delete_marker = version_info.delete_marker;
+                    if (!latest_ver_info && file_path === latest_version_path) {
+                        latest_ver_info = version_info;
+                    }
+                } else {
+                    const version_res = await this._delete_latest_version(fs_context, latest_version_path, { key, version_id });
+                    res.push(version_res);
+                    delete_marker_created = true;
+                }
+            } catch (err) {
+                res.push({ err_code: err.code, err_message: err.message });
+            }
+        }
+        // we try promote only if the latest version was deleted or we deleted a delete marker
+        // and no delete marker added (a new delete marker will be the latest - no need to promote)
+        if ((latest_ver_info || deleted_delete_marker) && !delete_marker_created) {
+            await this._promote_version_to_latest(fs_context, { key }, latest_ver_info);
+        }
+        // delete .versions/ if it's empty
+        const file_path = this._get_version_path(key);
+        await this._delete_path_dirs(file_path, fs_context);
+        return res;
+    }
+
+    // delete version_id - 
+    // 1. get version info, if it's empty - return
+    // 2. unlink key
+    // 3. promote second latest -> latest
+    async _delete_version_id(fs_context, file_path, params) {
+        const version_info = await this._get_version_info(fs_context, file_path);
+        if (!version_info) return {};
+        await nb_native().fs.unlink(fs_context, file_path);
+        await this._promote_version_to_latest(fs_context, params, version_info);
+        await this._delete_path_dirs(file_path, fs_context);
+        return {
+            deleted_delete_marker: version_info.delete_marker
+        };
+    }
+
+    // 1. if deleted version is not latest version and not a delete marker - skip
+    // 2. find max past version
+    //    2.1. if max_past_version does not exist / is a delete marker - skip, nothing to move
+    //    2.2. else - move max past version -> latest version path
+    // 3. if deleted version mtime < max_past_version mtime - skip (check if deleted version is latest or latest delete marker in .versions/)
+    // 4. move max past version -> latest version path
+    // conditions 1 and 2 guard on situations where we don't want to try move max version past to latest 
+    async _promote_version_to_latest(fs_context, params, deleted_version_info) {
+        dbg.log1('Namespace_fd._promote_version_to_latest', params, deleted_version_info);
+        const latest_ver_path = this._get_file_path({ key: params.key });
+        const deleted_latest = deleted_version_info && deleted_version_info.path === latest_ver_path;
+        // 1 - if deleted file is not a delete marker and not the latest - no need to promote max - return
```

Comment:
not sure if we still need 1 and 2 comments... but whatever you prefer

Reply:
let's keep it for now, we saw it's not easy to understand